library/network/src/lib/network/firewalld.rb

```diff
@@ -0,0 +1,324 @@
+
+    def initialize(option_str)
+      @option_str = option_str
+    end
+
+    def command
+      "#{COMMAND} #{option_str}".strip.squeeze(" ")
+    end
+
+    # Output resulting of executing the command
+    def output(need_output = true)
+      cmd_result = Yast::SCR.Execute(BASH_SCR_PATH, command)
+
+      # See firewall-cmd manpage for exit codes. Not all of them justify an
+      # exception. 0 and 1 can be used as true and false respectively.
+      case cmd_result["exit"]
+      when 0, 1
+        if need_output
+          cmd_result["stdout"]
+        else
+          log.debug "#{command} returned: #{cmd_result['stdout']}"
+          cmd_result["exit"].zero? ? true : false
+        end
+      else
+        raise FirewallCMDError, "Calling firewall-cmd (cmd: #{command}) failed: #{cmd_result["stderr"]}"
+      end
+    end
+  end
+
+  class FirewalldAPI
+    def self.create(type = :bash)
+      case type
+      when :bash
+        FirewalldBashAPI.new
+      when :dbus
+        nil
+      else
+        raise "Unsupported Firewalld API type: #{type}"
+      end
+    end
+  end
+
+  class FirewalldBashAPI
+
+    # Simple wrapper for commands. Returns true on success
+    def do_cmd(*args)
+      fwcmd = FWCmd.new(args.join(''))
+      fwcmd.output(false)
+    end
+
+    ### State ###
+
+    # True if firewalld is running
+    def is_running?
+      do_cmd("--state")
+    end
+
+    # True if firewalld was reloaded
+    def reload
+      do_cmd("--reload")
+    end
+
+    # True if firewalld was fully reloaded
+    def complete_reload
+      do_cmd("--complete-reload")
+    end
+
+    # True if runtime config has been made permanent
+    def make_permanent
+      do_cmd("--runtime-to-permanent")
+    end
+
+    ### Zones ####
+
+    # Return list of zones.
+    def get_zones
+      fwcmd = FWCmd.new("--permanent --get-zones")
+      fwcmd.output.split(" ")
+    end
+
+    # Return zone, or nil.
+    def get_zone_of_interface(iface)
+      do_cmd("--permanent --get-zone-of-interface=#{iface}")
+    end
+
+    # Return list of interfaces found in zone. Empty list if none found.
+    def get_interfaces_in_zone(zone)
+      fwcmd = FWCmd.new("--permanent --zone=#{zone} --list-interfaces")
+      fwcmd.output.split(" ")
+    end
+
+    # Return list of services that have been added to zone, or [].
+    def get_services_in_zone(zone)
+      fwcmd = FWCmd.new("--permanent --zone=#{zone} --list-services")
+      fwcmd.output.split(" ")
+    end
+
+    # Return list of ports that have been added to zone, or [].
+    def get_ports_in_zone(zone)
+      fwcmd = FWCmd.new("--permanent --zone=#{zone} --list-ports")
+      fwcmd.output.split(" ")
+    end
+
+    # Return list of protocols that have been added to zone, or [].
+    def get_protocols_in_zone(zone)
+      fwcmd = FWCmd.new("--permanent --zone=#{zone} --list-protocols")
+      fwcmd.output.split(" ")
+    end
+
+    # Return zone information
+    def list_all_zone(zone)
```

Comment:
"list_all_zone" sounds like a typo, but since "list_all_zones" exists too, we should find a better name.

Does it make sense to call firewall-cmd commands which all are named --list-foo and sometimes prefix them with "list" and sometimes not?